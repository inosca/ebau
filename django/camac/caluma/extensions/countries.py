COUNTRIES = {
    "Schweiz": "CH",
    "Deutschland": "DE",
    "Österreich": "AT",
    "Frankreich": "FR",
    "Italien": "IT",
    "Liechtenstein": "LI",
    "Ägypten": "EG",
    "Äquatorialguinea": "GQ",
    "Äthiopien": "ET",
    "Afghanistan": "AF",
    "Albanien": "AL",
    "Algerien": "DZ",
    "Amerikanisch-Samoa": "AS",
    "Amerikanische Jungferninseln": "VI",
    "Andorra": "AD",
    "Angola": "AO",
    "Anguilla": "AI",
    "Antarktis": "AQ",
    "Antigua und Barbuda": "AG",
    "Argentinien": "AR",
    "Armenien": "AM",
    "Aruba": "AW",
    "Aserbaidschan": "AZ",
    "Australien": "AU",
    "Bahamas": "BS",
    "Bahrain": "BH",
    "Bangladesch": "BD",
    "Barbados": "BB",
    "Belgien": "BE",
    "Belize": "BZ",
    "Benin": "BJ",
    "Bermuda": "BM",
    "Bhutan": "BT",
    "Bolivien": "BO",
    "Bosnien und Herzegowina": "BA",
    "Botsuana (Botswana)": "BW",
    "Brasilien": "BR",
    "Britische Jungferninseln": "VG",
    "Brunei": "BN",
    "Bulgarien": "BG",
    "Burkina Faso": "BF",
    "Burundi": "BI",
    "Caymaninseln, auch Kaimaninseln": "KY",
    "Chile": "CL",
    "China, Volksrepublik China": "CN",
    "Cookinseln": "CK",
    "Costa Rica": "CR",
    "Demokratische Republik Kongo": "CD",
    "Dominica": "DM",
    "Dominikanische Republik (Dom Rep)": "DO",
    "Dschibuti": "DJ",
    "Dänemark": "DK",
    "Ecuador": "EC",
    "El Salvador": "SV",
    "Elfenbeinküste": "CI",
    "Eritrea": "ER",
    "Estland": "EE",
    "Falklandinseln": "FK",
    "Fidschi": "FJ",
    "Finnland": "FI",
    "Französisch-Guayana": "GF",
    "Französisch-Polynesien": "PF",
    "Französische Süd- und Antarktisgebiete": "TF",
    "Färöer (Färöer-Inseln)": "FO",
    "Föderierte Staaten von Mikronesien": "FM",
    "Gabun": "GA",
    "Gambia": "GM",
    "Georgien": "GE",
    "Ghana": "GH",
    "Gibraltar": "GI",
    "Grenada": "GD",
    "Griechenland": "GR",
    "Grönland": "GL",
    "Guadeloupe": "GP",
    "Guam": "GU",
    "Guatemala": "GT",
    "Guinea": "GN",
    "Guinea-Bissau": "GW",
    "Guyana": "GY",
    "Haiti": "HT",
    "Honduras": "HN",
    "Hong Kong": "HK",
    "Indien": "IN",
    "Indonesien": "ID",
    "Irak": "IQ",
    "Iran (Persien)": "IR",
    "Irland": "IE",
    "Island": "IS",
    "Israel": "IL",
    "Jamaica": "JM",
    "Japan": "JP",
    "Jemen": "YE",
    "Jordanien": "JO",
    "Kambodscha": "KH",
    "Kamerun": "CM",
    "Kanada": "CA",
    "Kap Verde": "CV",
    "Kasachstan": "KZ",
    "Katar": "QA",
    "Kenia": "KE",
    "Kirgisistan (Kirgistan)": "KG",
    "Kiribati": "KI",
    "Kokosinseln, Keelinginseln": "CC",
    "Kolumbien": "CO",
    "Komoren": "KM",
    "Kosovo": "XK",
    "Kroatien": "HR",
    "Kuba": "CU",
    "Kuwait": "KW",
    "Laos": "LA",
    "Lesotho": "LS",
    "Lettland": "LV",
    "Libanon": "LB",
    "Liberia": "LR",
    "Libyen": "LY",
    "Litauen": "LT",
    "Luxemburg": "LU",
    "Macau": "MO",
    "Madagaskar": "MG",
    "Malawi": "MW",
    "Malaysia": "MY",
    "Malediven": "MV",
    "Mali": "ML",
    "Malta": "MT",
    "Marokko": "MA",
    "Marshallinseln": "MH",
    "Martinique": "MQ",
    "Mauritanien": "MR",
    "Mauritius": "MU",
    "Mayotte": "YT",
    "Mexico": "MX",
    "Monaco, Fürstentum Monaco": "MC",
    "Mongolei": "MN",
    "Montenegro": "ME",
    "Montserrat": "MS",
    "Mosambik": "MZ",
    "Myanmar, Birma, Burma": "MM",
    "Namibia": "NA",
    "Nauru": "NR",
    "Nepal": "NP",
    "Neukaledonien": "NC",
    "Neuseeland": "NZ",
    "Nicaragua": "NI",
    "Niederlande, Holland": "NL",
    "Niger": "NE",
    "Nigeria": "NG",
    "Nordkorea": "KP",
    "Nordmazedonien": "MK",
    "Norwegen": "NO",
    "Nördliche Marianen": "MP",
    "Oman": "OM",
    "Osttimor": "TL",
    "Pakistan": "PK",
    "Palau": "PW",
    "Panama": "PA",
    "Papua-Neuguinea": "PG",
    "Paraguay": "PY",
    "Peru": "PE",
    "Philippinen": "PH",
    "Pitcairn": "PN",
    "Polen": "PL",
    "Portugal": "PT",
    "Puerto Rico": "PR",
    "Republik China (Taiwan)": "TW",
    "Republik Kongo": "CG",
    "Republik Moldau, (Moldavien)": "MD",
    "Ruanda, Rwanda": "RW",
    "Rumänien": "RO",
    "Russland, Russische Föderation": "RU",
    "Réunion": "RE",
    "Salomonen": "SB",
    "Sambia": "ZM",
    "Samoa": "WS",
    "San Marino": "SM",
    "Saudi-Arabien": "SA",
    "Schweden": "SE",
    "Senegal": "SN",
    "Serbien": "RS",
    "Seychellen": "SC",
    "Sierra Leone": "SL",
    "Simbabwe": "ZW",
    "Singapur": "SG",
    "Slovenien": "SI",
    "Slowakei": "SK",
    "Somalia": "SO",
    "Spanien": "ES",
    "Sri Lanka (vorm. Ceylon)": "LK",
    "St. Kitts und Nevis": "KN",
    "St. Lucia": "LC",
    "St. Vincent und die Grenadinen": "VC",
    "Staat Palästina": "PS",
    "Sudan": "SD",
    "Suriname (Surinam)": "SR",
    "Swasiland (Eswatini)": "SZ",
    "Syrien": "SY",
    "São Tomé und Príncipe": "ST",
    "Südafrika": "ZA",
    "Südkorea (Republik Korea)": "KR",
    "Südsudan": "SS",
    "Tadschikistan": "TJ",
    "Tansania": "TZ",
    "Thailand": "TH",
    "Tibet": "CN",
    "Togo": "TG",
    "Tokelau": "TK",
    "Tonga": "TO",
    "Trinidad und Tobago": "TT",
    "Tschad": "TD",
    "Tschechien": "CZ",
    "Tunesien": "TN",
    "Turkmenistan": "TM",
    "Turks- und Caicosinseln": "TC",
    "Tuvalu": "TV",
    "Türkei": "TR",
    "Uganda": "UG",
    "Ukraine": "UA",
    "Ungarn": "HU",
    "Uruguay": "UY",
    "Usbekistan": "UZ",
    "Vanuatu": "VU",
    "Vatikanstadt (Heiliger Stuhl)": "VA",
    "Venezuela": "VE",
    "Vereinigte Arabische Emirate": "AE",
    "Vereinigte Staaten (Amerika, USA)": "US",
    "Vereinigtes Königreich (Großbritannien, England)": "GB",
    "Vietnam": "VN",
    "Wallis und Futuna": "WF",
    "Weihnachtsinsel": "CX",
    "Weißrussland": "BY",
    "Westsahara": "EH",
    "Zentralafrikanische Republik": "CF",
    "Zypern": "CY",
}
