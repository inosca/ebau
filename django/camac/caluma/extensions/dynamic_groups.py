import logging

from caluma.caluma_workflow.dynamic_groups import (
    BaseDynamicGroups,
    register_dynamic_group,
)

from camac.core.models import Activation, Circulation
from camac.instance.models import Instance
from camac.user.models import Service

log = logging.getLogger()


class CustomDynamicGroups(BaseDynamicGroups):
    def _get_activation(self, context):
        if not context:
            return Activation.objects.none()

        return Activation.objects.filter(pk=context.get("activation-id")).first()

    def _get_circulation(self, context):
        if not context:
            return Circulation.objects.none()

        return Circulation.objects.filter(pk=context.get("circulation-id")).first()

    def _get_responsible_service(self, case, filter_type, context):
        instance = (
            case.family.instance
            if hasattr(case.family, "instance")
            else Instance.objects.get(pk=context.get("instance"))
        )
        service = instance.responsible_service(filter_type=filter_type)

        if not service:
            log.error(f"No {filter_type} group found for instance {instance.pk}")

            return []

        return [str(service.pk)]

    @register_dynamic_group("municipality")
    def resolve_municipality(self, task, case, user, prev_work_item, context, **kwargs):
        return self._get_responsible_service(case, "municipality", context)

    @register_dynamic_group("construction_control")
    def resolve_construction_control(
        self, task, case, user, prev_work_item, context, **kwargs
    ):
        return self._get_responsible_service(case, "construction_control", context)

    @register_dynamic_group("circulation_service")
    def resolve_circulation_service(
        self, task, case, user, prev_work_item, context, **kwargs
    ):
        circulation = self._get_circulation(context)

        if not circulation:
            log.error("No service group found with the given context")
            return []

        return [str(circulation.service.pk)]

    @register_dynamic_group("activation_service")
    def resolve_activation_service(
        self, task, case, user, prev_work_item, context, **kwargs
    ):
        activation = self._get_activation(context)

        if not activation:
            log.error("No service group found with the given context")

            return []

        return [str(activation.service.pk)]

    @register_dynamic_group("activation_service_parent")
    def resolve_activation_service_parent(
        self, task, case, user, prev_work_item, context, **kwargs
    ):
        activation = self._get_activation(context)

        if not activation:
            log.error("No service_parent group found with the given context")

            return []

        return [str(activation.service_parent.pk)]

    @register_dynamic_group("distribution_create_inquiry")
    def resolve_distribution_create_inquiry(
        self, task, case, user, prev_work_item, context, **kwargs
    ):
        if context and context.get("addressed_groups") and prev_work_item:
            services = Service.objects.filter(
                pk__in=[
                    # Target service for the inquiry
                    *context.get("addressed_groups", []),
                    # Service that created an inquiry
                    *prev_work_item.addressed_groups,
                ],
                service_parent__isnull=True,  # Subservices can't create any inquiries
            )
            return [str(pk) for pk in services.values_list("pk", flat=True)]

        # If no context is given it's the first "create-inquiry" work item in
        # the distribution case which must be assigned to the municipality
        return self.resolve_municipality(
            task, case, user, prev_work_item, context, **kwargs
        )
