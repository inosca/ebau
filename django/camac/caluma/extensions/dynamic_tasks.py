from caluma.caluma_workflow.dynamic_tasks import BaseDynamicTasks, register_dynamic_task

from camac.instance.utils import should_continue_after_decision


class CustomDynamicTasks(BaseDynamicTasks):
    @register_dynamic_task("after-decision")
    def resolve_after_decision(self, case, user, prev_work_item, context):
        if case.workflow_id == "building-permit" and should_continue_after_decision(
            case.instance
        ):
            return ["sb1", "create-manual-workitems", "create-publication"]

        return []
