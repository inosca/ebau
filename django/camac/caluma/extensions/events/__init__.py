# Import all modules in this folder so the event receivers are registered
# automatically without having to define each file as caluma event receiver
# module

from . import (  # noqa: F401
    audit,
    decision,
    distribution,
    ebau_number,
    general,
    sb,
    simple_workflow,
)
