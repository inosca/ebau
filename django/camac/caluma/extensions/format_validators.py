from caluma.caluma_form.format_validators import BaseFormatValidator


class IntegerListFormatValidator(BaseFormatValidator):
    slug = "integer-list"
    name = {
        "en": "Comma separated list of integers",
        "de": "Komma separierte Liste von Integers",
    }
    regex = r"(^(\d+(,?|,\s?))+$)"
    error_msg = {
        "en": "Only comma separated intergers are permited",
        "de": "Nur Komma separierte Integers sind erlaubt",
    }
