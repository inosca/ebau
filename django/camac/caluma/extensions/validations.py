from caluma.caluma_core.validations import BaseValidation, validation_for
from caluma.caluma_form.models import Answer
from caluma.caluma_form.schema import SaveDocumentStringAnswer
from caluma.caluma_workflow.models import WorkItem
from caluma.caluma_workflow.schema import CompleteWorkItem
from django.conf import settings
from rest_framework import exceptions

from camac.caluma.utils import CamacRequest
from camac.ech0211.signals import file_subsequently
from camac.notification.utils import send_mail

CLAIM_QUESTION = "nfd-tabelle-table"
CLAIM_STATUS_QUESTION = "nfd-tabelle-status"
CLAIM_STATUS_IN_PROGRESS = "nfd-tabelle-status-in-bearbeitung"
CLAIM_STATUS_ANSWERED = "nfd-tabelle-status-beantwortet"

NOTIFICATION_CLAIM_IN_PROGRESS = "03-zusatzliche-unterlagen-notwendig-gesuchsteller"
NOTIFICATION_CLAIM_IN_PROGRESS_MUNICIPALITY = (
    "03-zusaetzliche-unterlagen-notwendig-gemeinde"
)
NOTIFICATION_CLAIM_ANSWERED = "03-nachforderung-beantwortet-leitbehorde"


class CustomValidation(BaseValidation):
    def _send_claim_notification(self, info, instance, template_slug, recipient_types):
        send_mail(
            template_slug,
            {"request": CamacRequest(info).request},
            recipient_types=recipient_types,
            instance={"type": "instances", "id": instance.pk},
        )

    def _send_claim_ech_event(self, info, instance):
        file_subsequently.send(
            sender=self.__class__,
            instance=instance,
            user_pk=None,  # Not needed, hence not querying for it
            group_pk=None,  # Not needed, hence not querying for it
        )

    @validation_for(SaveDocumentStringAnswer)
    def validate_save_document_string_answer(self, mutation, data, info):
        if data["question"].slug == CLAIM_STATUS_QUESTION:
            instance = data["document"].family.work_item.case.instance
            new_status = data["value"]

            try:
                old_status = (
                    data["document"].answers.get(question=CLAIM_STATUS_QUESTION).value
                )
            except Answer.DoesNotExist:
                old_status = None

            if old_status and new_status == old_status:
                # the status did not change, no further action
                return data

            if new_status == CLAIM_STATUS_IN_PROGRESS:
                # claim is now in progress, inform the applicant
                self._send_claim_notification(
                    info, instance, NOTIFICATION_CLAIM_IN_PROGRESS, ["applicant"]
                )
                self._send_claim_notification(
                    info,
                    instance,
                    NOTIFICATION_CLAIM_IN_PROGRESS_MUNICIPALITY,
                    ["inactive_municipality"],
                )

            if new_status == CLAIM_STATUS_ANSWERED:
                # claim is answered, inform the active service and create an
                # eCH event
                self._send_claim_notification(
                    info,
                    instance,
                    NOTIFICATION_CLAIM_ANSWERED,
                    ["leitbehoerde", "inactive_municipality"],
                )
                self._send_claim_ech_event(info, instance)

        return data

    @validation_for(CompleteWorkItem)
    def validate_complete_create_inquiry(self, mutation, data, info):
        work_item = WorkItem.objects.get(pk=data["id"])

        if (
            settings.DISTRIBUTION
            and work_item.task_id == settings.DISTRIBUTION["INQUIRY_CREATE_TASK"]
        ):
            service_id = str(info.context.user.group)
            addressed_groups = mutation.get_params(info)["input"]["context"][
                "addressed_groups"
            ]

            if service_id in addressed_groups:
                raise exceptions.ValidationError(
                    "Services can't create inquiries for themselves!"
                )

        return data
