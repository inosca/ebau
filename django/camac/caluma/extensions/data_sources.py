from caluma.caluma_data_source.data_sources import BaseDataSource
from caluma.caluma_data_source.utils import data_source_cache
from django.core.cache import cache
from django.utils.translation import gettext as _, override

from camac.core.models import Authority
from camac.user.models import Location, Service

from .countries import COUNTRIES


def get_municipality_label(service, municipality_prefix=False):
    translations = service.trans.all()
    label = {}

    for translation in translations:
        name = (
            translation.name.replace(
                "Leitbehörde", "Gemeinde" if municipality_prefix else ""
            ).replace(
                "Autorité directrice", "Municipalité" if municipality_prefix else ""
            )
        ).strip()

        if service.disabled:
            with override(translation.language):
                postfix = _("not activated")
                text = f"{name} ({postfix})"
        else:
            text = name

        label[translation.language] = text

    for language in ["de", "fr"]:
        if language not in label.keys():
            label[language] = list(label.values())[0]

    return label


def get_others_option():
    label = {}

    for language in ["de", "fr"]:
        with override(language):
            label[language] = _("Others")

    return ["-1", label]


class Municipalities(BaseDataSource):
    info = "List of municipalities from Camac"

    def get_data(self, user, question, context):
        cache_key = f"data_source_{type(self).__name__}"
        include_disabled = (
            hasattr(user, "group")
            and Service.objects.filter(
                pk=user.group, service_group__name="district"
            ).exists()
            or (hasattr(user, "camac_role") and user.camac_role == "support")
        )

        if include_disabled:
            cache_key += "_with_disabled"
            filters = {}
        else:
            filters = {"disabled": False}

        return cache.get_or_set(cache_key, lambda: self._get_data(filters), 3600)

    def _get_data(self, filters):
        services = (
            Service.objects.select_related("service_group")
            .filter(
                service_parent__isnull=True,
                service_group__name="municipality",
                **filters,
            )
            .prefetch_related("trans")
        )

        return sorted(
            [[service.pk, get_municipality_label(service)] for service in services],
            key=lambda x: x[1]["de"].casefold(),
        )


class Locations(BaseDataSource):
    info = "List of locations from Camac"

    def get_data(self, user, question, context):
        cache_key = f"data_source_{type(self).__name__}"
        include_special = (
            hasattr(user, "camac_role") and user.camac_role != "Portal User"
        )

        if include_special:
            cache_key += "_with_special"
            filters = {}
        else:
            filters = {"zip__isnull": False}

        return cache.get_or_set(cache_key, lambda: self._get_data(filters), 3600)

    def _get_data(self, filters):
        locations = Location.objects.filter(**filters)

        return sorted(
            [
                [int(loc.communal_federal_number), loc.name]
                for loc in locations.iterator()
            ],
            key=lambda x: x[1].casefold(),
        )


form_mapping_by_koor = {
    "Koordinationsstelle Baugesuche BG": [
        [42, "Internes Mitberichtsverfahren / Genehmigungsverfahren"],
        [46, "Mitberichtverfahren zu Plangenehmigungsverfahren Militär (PGV Militär)"],
        [
            45,
            "Mitberichtverfahren zu Plangenehmigungsverfahren Luftfahrt (PGV Luftfahrt)",
        ],
        [
            250,
            "Mitberichtverfahren zu Plangenehmigungsverfahren Seilbahn (PGV Seilbahn)",
        ],
    ],
    "Koordinationsstelle Baudirektion BD": [
        [223, "PGV öffentliche Gewässer (Artikel 12 WBG)"],
        [224, "PGV private Gewässer (Artikel 19 WBG)"],
        [181, "PGV Kantonsstrasse (Artikel 30 StrG)"],
        [201, "PGV Gemeindestrasse (Artikel 30 StrG)"],
        [221, "PGV Korporationsstrasse (Artikel 30 StrG)"],
        [222, "PGV vereinfachtes Verfahren (Artikel 31 StrG)"],
        [
            225,
            "Konzessionsverfahren Regierungsrat für Anlagen bis 1000 kW (Artikel 18 GNG)",
        ],
        [241, "Konzessionsverfahren Landrat für Anlagen ab 1000 kW (Artikel 18 GNG)"],
        [242, "Konzessionsverfahren Baudirektion für Wärmeentnahmen (Artikel 40 GNG)"],
        [243, "Mitberichtsverfahren / PGV nach Starkstromverordnung (EleG)"],
        [244, "Mitberichtsverfahren / PGV nach Eisenbahngesetz (EBG)"],
        [245, "Mitberichtsverfahren / PGV nach Nationalstrassengesetz (NSG)"],
        [
            246,
            "Konzessionsverfahren Baudirektion Beanspruchung Kantonsgewässer bis 500 m2 (Artikel 3 ORR)",
        ],
        [248, "Internes Mitberichtsverfahren BD"],
        [286, "Land- und Rechtserweb"],
    ],
    "Koordinationsstelle Nutzungsplanung NP": [
        [161, "Internes Mitberichtsverfahren"],
        [46, "Mitberichtverfahren zu Plangenehmigungsverfahren Militär (PGV Militär)"],
        [
            45,
            "Mitberichtverfahren zu Plangenehmigungsverfahren Luftfahrt (PGV Luftfahrt)",
        ],
        [
            250,
            "Mitberichtverfahren zu Plangenehmigungsverfahren Seilbahn (PGV Seilbahn)",
        ],
    ],
    "Koordinationsstelle Energie AfE": [
        [256, "Internes Mitberichtsverfahren AfE"],
        [
            257,
            "Konzessionsverfahren Regierungsrat für Anlagen bis 1000 kW (Artikel 18 GNG)",
        ],
        [258, "Konzessionsverfahren Landrat für Anlagen ab 1000 kW (Artikel 18 GNG)"],
        [259, "Konzessionsverfahren Baudirektion für Wärmeentnahmen (Artikel 40 GNG)"],
        [
            289,
            "Konzessionsverfahren Korporationsgewässer mit Genehmigung durch Regierungsrat (Artikel 15 GNG)",
        ],
    ],
    "Koordinationsstelle Forst und Jagd AFJ": [
        [260, "Internes Mitberichtsverfahren AFJ"],
    ],
    "Koordinationsstelle Landwirtschaft ALA": [
        [254, "Internes Mitberichtsverfahren ALA"],
        [
            255,
            "Internes Mitberichtsverfahren ALA (Korporationsstrassen mit Subventionierung)",
        ],
    ],
    "Koordinationsstelle Sicherheitsdirektion SD": [
        [287, "Internes Mitberichtsverfahren SD AfKP"],
        [288, "Internes Mitberichtsverfahren SD"],
    ],
    "Koordinationsstelle Umweltschutz AfU": [
        [251, "Internes Mitberichtsverfahren AfU"],
        [252, "Internes Mitberichtsverfahren nach Artikel 30 BG Umweltschutz"],
        [253, "Internes Mitberichtsverfahren / Bewilligungsverfahren gemäss UVP"],
    ],
}


class Mitberichtsverfahren(BaseDataSource):
    info = "List of different types of 'Mitberichtsverfahren' (role-dependent)"

    def get_data(self, user, question, context):
        if not hasattr(user, "camac_role"):  # pragma: no cover
            return []
        return form_mapping_by_koor.get(user.camac_role, [])


class Services(BaseDataSource):
    info = "List of services, municipalities and RSTAs from Camac"

    @data_source_cache(timeout=3600)
    def get_data(self, user, question, context):
        services = (
            Service.objects.select_related("service_group")
            .filter(
                service_parent__isnull=True,
                service_group__name__in=[
                    "service",
                    "municipality",
                    "district",
                ],
                disabled=False,
            )
            .prefetch_related("trans")
        )

        data = sorted(
            [
                [str(service.pk), get_municipality_label(service, True)]
                for service in services
            ],
            key=lambda x: x[1]["de"].casefold(),
        ) + [get_others_option()]

        return data


class Countries(BaseDataSource):
    info = "List of all countries in the world with opinionated sorting"

    @data_source_cache(timeout=3600)
    def get_data(self, user, question, context):
        return COUNTRIES


class Authorities(BaseDataSource):
    info = "List of authorities from camac"

    def get_data(self, user, question, context):
        return [[authority.pk, authority.name] for authority in Authority.objects.all()]
