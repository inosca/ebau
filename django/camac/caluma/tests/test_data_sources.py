from collections import namedtuple

import pytest

from ..extensions.countries import COUNTRIES
from ..extensions.data_sources import (
    Authorities,
    Countries,
    Locations,
    Mitberichtsverfahren,
    Municipalities,
    Services,
)


@pytest.mark.parametrize(
    "role,expected_count", [("Portal User", 1), ("Some internal role", 2)]
)
def test_locations(db, role, location_factory, expected_count):
    User = namedtuple("OIDCUser", "camac_role")
    user = User(camac_role=role)

    location_factory(name="Foo", zip=123)
    location_factory(name="Foo", zip=None)

    data = Locations().get_data(user, None, None)
    assert len(data) == expected_count


@pytest.mark.parametrize(
    "role,expected_count",
    [("Koordinationsstelle Baugesuche BG", 4), ("Something else", 0)],
)
def test_mitberichtsverfahren(db, role, location_factory, expected_count):
    User = namedtuple("OIDCUser", "camac_role")
    user = User(camac_role=role)

    location_factory(name="Foo", zip=123)
    location_factory(name="Foo", zip=None)

    data = Mitberichtsverfahren().get_data(user, None, None)
    assert len(data) == expected_count


@pytest.mark.parametrize(
    "test_class,expected,is_rsta",
    [
        (Authorities, [[1, "Baukommission Altdorf"]], False),
        (Municipalities, [[1, {"de": "Bern", "fr": "Berne"}]], False),
        (
            Municipalities,
            [[2, {"de": "Biel (nicht aktiviert)", "fr": "Bienne (non activé)"}]],
            True,
        ),
        (
            Services,
            [
                ["1", {"de": "Gemeinde Bern", "fr": "Municipalité Berne"}],
                ["3", {"de": "service3", "fr": "service3"}],
                ["4", {"de": "service4", "fr": "service4"}],
                ["-1", {"de": "Andere", "fr": "Autres"}],
            ],
            False,
        ),
        (
            Countries,
            COUNTRIES,
            False,
        ),
    ],
)
def test_data_sources(
    db,
    multilang,
    service_factory,
    service_t_factory,
    service_group_factory,
    test_class,
    expected,
    is_rsta,
    authority_factory,
):
    if is_rsta:
        service1 = service_factory(
            pk=1,
            trans__name="service1",
            trans__language="de",
            disabled=False,
            service_group__name="district",
        )
    else:
        service1 = service_factory(
            pk=1,
            trans__name="Leitbehörde Bern",
            trans__language="de",
            disabled=False,
            service_group__name="municipality",
        )
        service_t_factory(
            service=service1, name="Autorité directrice Berne", language="fr"
        )
        authority_factory(pk=1, name="Baukommission Altdorf")

    service2 = service_factory(
        pk=2,
        trans__name="Leitbehörde Biel",
        trans__language="de",
        disabled=True,
        service_group__name="municipality",
    )
    service_t_factory(
        service=service2, name="Autorité directrice Bienne", language="fr"
    )

    service_factory(
        pk=3,
        trans__name="service3",
        trans__language="de",
        disabled=False,
        service_group__name="district",
    )

    service_factory(
        pk=4,
        trans__name="service4",
        trans__language="de",
        disabled=False,
        service_group__name="service",
    )

    User = namedtuple("OIDCUser", "group")
    user = User(group=service1.pk)

    data = test_class().get_data(user, None, None)

    assert data == expected
