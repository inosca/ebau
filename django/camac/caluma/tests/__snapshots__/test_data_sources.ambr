# serializer version: 1
# name: test_landowners_so
  list([
    'Powell, Lee and Jordan, Linda White',
    'Cynthia Hester',
    'Smith Group, Rodney Maldonado',
    'Austin Evans',
  ])
# ---
